### ISSUE TITLE
Polygon validation results in 'GeometryCollection' object has no attribute 'exterior'

### ISSUE THREAD

**Issue by chododom** [srai-lab/srai#309]
I am trying to download the features for a larger area (specifically the country of Germany) using the OSMPbfLoader.
However, I am running into problems, because when I run this code:
```
area = geocode_to_region_gdf("Germany")
regionalizer = H3Regionalizer(resolution=6)

loader = OSMPbfLoader()
features = loader.load(area, HEX2VEC_FILTER)
```
it results in this error:
```
srai/loaders/osm_loaders/pbf_file_downloader.py:249, in PbfFileDownloader._simplify_polygon(self, polygon)
    239     simplified_polygon = (
    240         tp.Topology(
    241             polygon,
   (...)
    246         .geometry[0]
    247     )
    248     simplified_polygon = make_valid(simplified_polygon)
--> 249     if len(simplified_polygon.exterior.coords) < 1000:
    250         break
    252 if len(simplified_polygon.exterior.coords) > 1000:

AttributeError: 'GeometryCollection' object has no attribute 'exterior'
```

I have backtracked the issue down to line `simplified_polygon = make_valid(simplified_polygon)`. If I comment this line out, then the function proceeds.

However, when I do this, the process ends up failing anyway as the download request throws the error:
```
[UserWarning: {"error": "Too many nodes: 390103584.0. Maximum: 100000000"}]
```

Is there a way around this for larger map areas? Can I somehow split the download and concatenate the results?

Thanks!

**Comment by simonusher** [comment]
Hi! Thanks for reporting that. @RaczeQ would you be able to take a look?

**Comment by RaczeQ** [comment]
Hello 😄 

I've tracked down this issue (after commenting out a `make_valid` command as you've pointed out) and it seems like the Protomaps service is limiting such big extracts (creating pbf files with up to 100000000 nodes).

We didn't expect anyone to be using this function to download data for all countries, but at the same time, we weren't aware of this Protomaps limitation.

Fortunately, we are prepared for that and we can load existing pbf files from Geofabrik for example. We are even creating a function to automatically find a matching geofabrik extract and download it (probably will be available this or next month).

For now, here is the snippet that will work for you (it will use some classes that are not described yet in examples):
```python
from srai.utils import download_file, geocode_to_region_gdf
from srai.loaders import OSMPbfLoader
from srai.loaders.osm_loaders.filters import HEX2VEC_FILTER
from srai.loaders.osm_loaders.pbf_file_handler import PbfFileHandler

area = geocode_to_region_gdf("Germany")

germany_extract_pbf_file_path = 'germany-latest.osm.pbf'

# Look here for more available extracts: http://download.geofabrik.de/europe/germany.html

# extract_url = 'http://download.geofabrik.de/europe/germany-latest.osm.pbf'
extract_url = 'http://download.geofabrik.de/europe/germany/berlin-latest.osm.pbf' # test on smaller scale
download_file(extract_url, germany_extract_pbf_file_path)

features = PbfFileHandler(tags=HEX2VEC_FILTER).get_features_gdf(file_paths=[germany_extract_pbf_file_path])
```

**Comment by chododom** [comment]
Hi @RaczeQ , thank you very much, this is a good solution.

On a side note unrelated to the bug report - do you happen to know if there is a way of processing such large numbers of nodes more efficiently? The map of whole Germany seems to be too big for the loader to extract the features in reasonable time, I was wondering if it is possible to use several smaller files or somehow reduce the information of the OpenStreetMap format to only specific tags (for my usecase I actually only need information about nature and landuse and such).

**Comment by RaczeQ** [comment]
For now, we don't have this feature (and it will be hard to integrate with Python). We are working on a faster loading using GDAL driver within duckdb. Unfortunately, `pyosmium` dependency that is currently in our library is quite slow since it's Python binding.

If you want to reduce the size of a file / exclude tags etc, I'd encourage you to preprocess the pbf file with external C/C++ tools such as osmium (https://docs.osmcode.org/osmium/latest/osmium-tags-filter.html). You can also use osmium to clip a pbf file using a bounding box or a required polygon.

**Comment by RaczeQ** [comment]
If you have any more questions about working with PBF files, you can continue asking them here. I've closed this issue with the linked PR for now.

**Comment by chododom** [comment]
Hi, thank you very much for the tip, I ended up using the mentioned osmium tool for filtering out tags that were not related to my use case, which reduced the size quite a bit.

However, I am now hitting a point in the computation, where even that becomes a problem. When I try to load the features of the filtered map of Germany:
```
area = geocode_to_region_gdf("Germany")
loader = OSMPbfLoader(pbf_file='./files/germany-filtered.osm')
features = loader.load(area, HEX2VEC_FILTER)
```

it runs well for the first roughly 88 %, but then it drastically slows down. Is there a part of the processing workload that is much more computationally heavy at the end? The first 88 % ran for about 20 minutes, now the estimate for the rest of the computation is about 60 hours.

```
[germany-filtered.osm] Counting pbf features: 84573655it [04:37, 304992.69it/s]
[germany-filtered.osm] Parsing pbf file #1:  88%|█████████████████████▏  | 74810742/84573655 [25:55<52:19:03, 51.84it/s]
```

The same issue happens if I use osmium to split the osm.pbf file into several smaller ones (I tried splitting them into files where max node count is 1 million), the feature extraction ran to 88 % in a couple of seconds, but then it got incredibly stuck.

Any ideas why this could be and how it could be solved?
